cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/Node/LedgerEvents/Governance/ProposeNewConstitution.hs

```diff
@@ -0,0 +1,414 @@
+    , "--out-file", voteTxBodyFp
+    ]
+  let _wallet1VKeyFp = paymentVKey . paymentKeyInfoPair $ head  wallets
+      _wallet2VKeyFp = paymentVKey . paymentKeyInfoPair $ wallets !! 1
+      _wallet3VKeyFp = paymentVKey . paymentKeyInfoPair $ wallets !! 2
+
+
+  void $ H.execCli' execConfig
+    [ "conway", "transaction", "sign"
+    , "--testnet-magic", show @Int testnetMagic
+    , "--tx-body-file", voteTxBodyFp
+    , "--signing-key-file", paymentSKey $ paymentKeyInfoPair $ head wallets
+    , "--signing-key-file", drepSKeyFp 1
+    , "--signing-key-file", drepSKeyFp 2
+    , "--signing-key-file", drepSKeyFp 3
+    , "--out-file", voteTxFp
+    ]
+
+  void $ H.execCli' execConfig
+    [ "conway", "transaction", "submit"
+    , "--testnet-magic", show @Int testnetMagic
+    , "--tx-file", voteTxFp
+    ]
+
+  -- We check that constitution was succcessfully ratified
+
+  !eConstitutionAdopted
+    <- runExceptT $ handleIOExceptT IOE
+                  $ runExceptT $ foldBlocks
+                      (File $ configurationFile testnetRuntime)
+                      (File socketPath)
+                      FullValidation
+                      [] -- Initial accumulator state
+                      (foldBlocksCheckConstitutionWasRatified constitutionHash)
+
+
+  case eConstitutionAdopted of
+    Left (IOE e) ->
+      H.failMessage callStack
+        $ "foldBlocksCheckConstitutionWasRatified failed with: " <> show e
+    Right (Left e) ->
+      H.failMessage callStack
+        $ "foldBlocksCheckConstitutionWasRatified failed with: " <> Text.unpack (renderFoldBlocksError e)
+    Right (Right _events) -> success
+
+foldBlocksCheckProposalWasSubmitted
+  :: TxId -- TxId of submitted tx
+  -> Env
+  -> LedgerState
+  -> [LedgerEvent]
+  -> BlockInMode -- Block i
+  -> [LedgerEvent] -- ^ Accumulator at block i - 1
+  -> IO ([LedgerEvent], FoldStatus) -- ^ Accumulator at block i and fold status
+foldBlocksCheckProposalWasSubmitted txid _ _ allEvents _ _ =
+  if any (filterNewGovProposals txid) allEvents
+  then return (allEvents , StopFold)
+  else return ([], ContinueFold)
+
+retrieveGoveranceActionIndex
+  :: (HasCallStack, MonadTest m)
+  => TxId -> [LedgerEvent] -> m Word32
+retrieveGoveranceActionIndex txid events = do
+  let newGovProposals = filter (filterNewGovProposals txid) events
+  if null newGovProposals
+  then H.failMessage callStack "retrieveGoveranceActionIndex: No new governance proposals found"
+  else
+    -- In this test there will only be one
+    case head newGovProposals of
+      NewGovernanceProposals _ (AnyProposals props) -> do
+        let govActionStates = [i
+                              | Ledger.GovActionIx i <- map Ledger.gaidGovActionIx . Map.keys $ Ledger.proposalsGovActionStates props
+                              ]
+        return $ head  govActionStates
+      other -> H.failMessage callStack $ "retrieveGoveranceActionIndex: Expected NewGovernanceProposals, got: " <> show other
```

Comment:
```suggestion
  case newGovProposals of
    [] -> H.failMessage callStack "retrieveGoveranceActionIndex: No new governance proposals found"
    NewGovernanceProposals _ (AnyProposals props) : _ -> do
          let govActionStates = [i
                                | Ledger.GovActionIx i <- map Ledger.gaidGovActionIx . Map.keys $ Ledger.proposalsGovActionStates props
                                ]
          return $ head  govActionStates
    other -> H.failMessage callStack $ "retrieveGoveranceActionIndex: Expected NewGovernanceProposals, got: " <> show other
```

Reply:
This entire function was refactored based on a comment from John. Lmk what you think.

Reply:
LGTM :heavy_check_mark: